# K Nearest  Neighbour Classifier

In [1]:
# importing libraries:
import pandas as pd 
import numpy as np 
import seaborn as sns
import matplotlib.pyplot as plt 
%matplotlib inline

In [2]:
 # creating the dataset for KNN classification by using make_classification
from sklearn.datasets import make_classification

x, y = make_classification(
    n_samples= 1000,  # Observations
    n_features= 3,     # we take 3 features for classification
    n_redundant= 1, 
    n_classes= 2 ,  # Binary target
    random_state= 999
)

In [3]:
# train test split 
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size= 0.33, random_state= 42)

In [5]:
# importing KN classifier
from sklearn.neighbors import KNeighborsClassifier

In [6]:
classfier = KNeighborsClassifier(n_neighbors= 5, algorithm= 'auto')
classfier.fit(x_train, y_train)      # applying fit 

KNeighborsClassifier()

In [7]:
y_pred = classfier.predict(x_test)

In [8]:
# getting output in [confusion_metrix, accuracy score, classification report]
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix


print(accuracy_score(y_pred, y_test))
print(confusion_matrix(y_pred, y_test))
print(classification_report(y_pred, y_test))

0.906060606060606
[[158  20]
 [ 11 141]]
              precision    recall  f1-score   support

           0       0.93      0.89      0.91       178
           1       0.88      0.93      0.90       152

    accuracy                           0.91       330
   macro avg       0.91      0.91      0.91       330
weighted avg       0.91      0.91      0.91       330



# Internal Assignment Given by @Krish Naik sir

By using value k = [1,2,3,4,5,6,7,8,9,10] get best Accuracy for the model by using GridSearchCV and RandomizedsearchCV

In [9]:
# GridSearchCV

from sklearn.model_selection import GridSearchCV

In [11]:
param_grid = {'n_neighbors': [1,2,3,4,5,6,7,8,9,10]}   # defining param_grid to search
grid_search = GridSearchCV(classfier, param_grid, cv = 5, scoring= 'accuracy')  ##creating Gridsearchcv object
grid_search.fit(x_train, y_train)  ## fit

#geting the best Accuracy score
best_parameter = grid_search.best_params_
best_accuracy = grid_search.best_score_
print('Best Parameter for k is:', best_parameter)
print('Best Accuracy Score is:', best_accuracy)

Best Parameter for k is: {'n_neighbors': 9}
Best Accuracy Score is: 0.9029850746268657


In [12]:
 # with best parameter get best accuracy of the model 
classfier = KNeighborsClassifier(n_neighbors= 9, algorithm= 'auto')
classfier.fit(x_train, y_train) 

KNeighborsClassifier(n_neighbors=9)

In [13]:
y_pred = classfier.predict(x_test)

In [15]:
print('Best Accuracy Score in (GridSearch)is :',accuracy_score(y_pred, y_test))

Best Accuracy Score in (GridSearch)is : 0.9121212121212121


In [17]:
# RandomizedSeachCV

from sklearn.model_selection import RandomizedSearchCV

In [18]:
param_dist = {'n_neighbors' : np.arange(1,11)}  # difinging parameter distribution to sample form
random_search = RandomizedSearchCV(classfier, param_distributions= param_dist, n_iter= 10, cv = 5, 
                                   scoring= 'accuracy', random_state= 42)
random_search.fit(x_train, y_train)

#geting best Accuracy;
best_params = random_search.best_params_
best_random_accuracy = random_search.best_score_
print('Best parameter of k is:', best_params)
print('Best_Accuracy-Score is :', best_accuracy)

Best parameter of k is: {'n_neighbors': 9}
Best_Accuracy-Score is : 0.9029850746268657


# Obeservation:
by using k = 9 in GridsearchCV we got  **91 %** accuracy and by using Randomized search Cv we also got **90%** accuracy.